### PR TITLE
feat: handle proprely all possible union and optional of tensor in da stack

### PR DESCRIPTION
**Pull request by samsja**
# Context

* refactor code to use is tensor union and handle more Union of tensor types 
* rename Embedding to AnyEmbedding
* rename `util` to `utils`

### CODE REVIEW

File: docarray/array/array_stacked.py [hunk 1]
```diff
@@ -97,16 +94,13 @@ def _from_columns(
     def _create_columns(
         cls: Type[T], docs: DocumentArray, tensor_type: Type['AbstractTensor']
     ) -> Dict[str, Union['TorchTensor', T, NdArray]]:
+        from docarray.util._typing import is_tensor_union
```
Comment (JohannesMessner) [inline docarray/array/array_stacked.py:97]: Why the lazy import?

File: docarray/typing/tensor/abstract_tensor.py [hunk 2]
```diff
@@ -117,3 +117,8 @@ def __docarray_from_native__(cls: Type[T], value: Any) -> T:
     def get_comp_backend() -> Type[AbstractComputationalBackend]:
         """The computational backend compatible with this tensor type."""
         ...
+
+    @staticmethod
+    def __docarray_normalize_value__(val: Any) -> Any:
+        """Normalize a value to be compatible with this tensor type."""
+        return val
```
Comment (JohannesMessner) [inline docarray/typing/tensor/abstract_tensor.py:124]: Side note: I know we do this in other places too, but I think the Python convention is that double-dunder `__cool_method__` methods are methods that are called "magically" under the hood, but can be overwritten.
So maybe this should be called just `docarray_normalize_value()`?

File: docarray/typing/tensor/abstract_tensor.py [hunk 3]
```diff
@@ -117,3 +117,8 @@ def __docarray_from_native__(cls: Type[T], value: Any) -> T:
     def get_comp_backend() -> Type[AbstractComputationalBackend]:
         """The computational backend compatible with this tensor type."""
         ...
+
+    @staticmethod
+    def __docarray_normalize_value__(val: Any) -> Any:
```
Comment (JohannesMessner) [inline docarray/typing/tensor/abstract_tensor.py:122]: It is not clear to me what this method is intended for, even after reading the implementation for TorchTensor. Maybe we can a bit more info to the docstring here?

**Review by JohannesMessner — CHANGES_REQUESTED** [review]

File: docarray/array/array_stacked.py (hunk 1 above)
Comment (samsja) [reply docarray/array/array_stacked.py:97]: ah I need to fix. I had circular import so I did this

File: docarray/typing/tensor/abstract_tensor.py (hunk 3 above)
Comment (samsja) [reply docarray/typing/tensor/abstract_tensor.py:122]: sure

File: docarray/typing/tensor/abstract_tensor.py (hunk 2 above)
Comment (samsja) [reply docarray/typing/tensor/abstract_tensor.py]: I refactor the code. I don't have this normalize like this anymore

File: docarray/typing/tensor/abstract_tensor.py (hunk 3 above)
Comment (samsja) [reply docarray/typing/tensor/abstract_tensor.py]: this is refactor

File: docarray/util/_typing.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from typing import Any, get_args
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/util/_typing.py:1]: well I want to package to be private. Nobody should use this function outside of us. You prefer to have everything private ? I was assuming if the package is private then we are good

File: docarray/computation/abstract_comp_backend.py [hunk 5]
```diff
@@ -31,6 +31,13 @@ def stack(
     def n_dim(array: 'TTensor') -> int:
         ...
 
+    @staticmethod
+    @abstractmethod
+    def none_value() -> typing.Any:
+        """Provide a compatible value that represents None in the Tensor Backend."""
+        ...
+        # return torch.tensor(float('nan'))
```
Comment (JohannesMessner) [inline docarray/computation/abstract_comp_backend.py:39]: comment

File: docarray/computation/numpy_backend.py [hunk 6]
```diff
@@ -44,6 +44,11 @@ def stack(
     def n_dim(array: 'np.ndarray') -> int:
         return array.ndim
 
+    @staticmethod
+    def none_value() -> Any:
+        """Provide a compatible value that represents None in torch."""
+        return None
```
Comment (JohannesMessner) [inline docarray/computation/numpy_backend.py:50]: Why not `np.array(float('nan'))`?

**Review by JohannesMessner — CHANGES_REQUESTED** [review]

File: docarray/computation/numpy_backend.py (hunk 6 above)
Comment (samsja) [reply docarray/computation/numpy_backend.py:50]: Numpy do handle None properly. Though we can decide to use Nan as well to do the same as in torch not sure though that it makes sense

**Review by JohannesMessner — APPROVED** [review]
